### PR TITLE
Fixed issue #18338: For participants, expiration date can be set before start date

### PR DESCRIPTION
**Pull request by gabrieljenik**
- Moving core validation function to adminbasics (alos loaded in POS:BEGIN :) )
- Implementing it or the token form

### CODE REVIEW

File: assets/packages/adminbasics/build/adminbasics.js [hunk 1]
```diff
@@ -29781,6 +29781,31 @@
         if (e.target != this) return;
         $('#accordion').find('.panel-collapse.collapse').not('#' + $(this).attr('id')).collapse('hide');
       });
+    },
+
+    /**
+     * Validates that an end date is not lower than a start date
+     */
+    validateEndDateHigherThanStart: function validateEndDateHigherThanStart(startDatePicker, endDatePicker, errorCallback) {
```
Comment (olleharstedt) [inline assets/packages/adminbasics/build/adminbasics.js:29789]: Hm, isn't this code duplication from a previous commit?

Comment (olleharstedt) [reply assets/packages/adminbasics/build/adminbasics.js:29789]: Oh, you removed it. But you still have it in two places in this PR?

**Review by olleharstedt — CHANGES_REQUESTED** [review]

File: assets/packages/adminbasics/build/adminbasics.js (hunk 1 above)
Comment (Shnoulle) [reply assets/packages/adminbasics/build/adminbasics.js:29789]: Oh right : https://github.com/LimeSurvey/LimeSurvey/pull/2679/files#diff-ede93d7c4e63dba99f44a55b638edd64610bd2096268abc5375ba0bc21ffa149R119

2nd place

File: assets/packages/adminbasics/build/adminbasics.js [hunk 2]
```diff
@@ -29781,6 +29781,31 @@
         if (e.target != this) return;
         $('#accordion').find('.panel-collapse.collapse').not('#' + $(this).attr('id')).collapse('hide');
       });
+    },
+
+    /**
+     * Validates that an end date is not lower than a start date
+     */
```
Comment (Shnoulle) [inline assets/packages/adminbasics/build/adminbasics.js:29788]: More doc here ? What muts be each params ?

File: assets/packages/adminbasics/build/adminbasics.js [hunk 3]
```diff
@@ -29781,6 +29781,31 @@
...
+    validateEndDateHigherThanStart: function validateEndDateHigherThanStart(startDatePicker, endDatePicker, errorCallback) {
+      if (!startDatePicker || !startDatePicker.date()) {
```
Comment (Shnoulle) [inline assets/packages/adminbasics/build/adminbasics.js:29790]: if startDatePicker didn't have date() function : it broke validation and form can not be submitted.
no ?

File: assets/packages/adminbasics/src/parts/globalMethods.js [hunk 4]
```diff
@@ -112,7 +112,26 @@ const globalWindowMethods = {
             if(e.target != this) return;
             $('#accordion').find('.panel-collapse.collapse').not('#'+$(this).attr('id')).collapse('hide');
         });
-    }
+    },
+    /**
+     * Validates that an end date is not lower than a start date
+     */
+    validateEndDateHigherThanStart: (startDatePicker, endDatePicker, errorCallback) => {
```
Comment (Shnoulle) [inline assets/packages/adminbasics/src/parts/globalMethods.js:119]: duplicate ;)

**Review by Shnoulle — CHANGES_REQUESTED** [review]
This need to remove duplicate (thank Olle ;).

And clearly : need more doc on params

**Comment by Shnoulle** [comment]
An alternative way i like to use is to set [setCustomValidity](https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/setCustomValidity)  on one (or both) date element after it's updated.

Then : when you try to submit : focus is set to invalid input directly.

**Comment by olleharstedt** [comment]
@gabrieljenik Can you remove the code duplication in this PR, please?

**Comment by gabrieljenik** [comment]
> @gabrieljenik Can you remove the code duplication in this PR, please?

@olleharstedt There is no duplication I believe.
One piece of code is the source (globalMethods.js). The other one is its build.

File: assets/packages/adminbasics/src/parts/globalMethods.js (hunk 4 above)
Comment (olleharstedt) [reply assets/packages/adminbasics/src/parts/globalMethods.js:122]: This is not a build, it's in src/ folder. So still duplicate. I think.

**Comment by olleharstedt** [comment]
> One piece of code is the source (globalMethods.js). The other one is its build.

I don't think it's in build/, looks like two copies in src/ folder.

File: assets/packages/adminbasics/src/parts/globalMethods.js (hunk 4 above)
Comment (olleharstedt) [reply assets/packages/adminbasics/src/parts/globalMethods.js:122]: Sorry, I'm wrong.

**Comment by olleharstedt** [comment]
Nope, I'm wrong.

**Comment by gabrieljenik** [comment]
No worries. Thanks!

**Pull request merged**